### DISCUSSION TITLE
Alpine ignores events triggered by JQuery

### DISCUSSION THREAD

**Discussion by MartinMystikJonas** [alpinejs/alpine#2062] [1. Help] [answered]
I just spent few hours solving strange issue.  We use jQuery plugin to decorate number input fields with nice [+] and [-] buttons. These buttons changed value of input with x-model but Apline did not reacted to it. Apline ignored events triggered by jQuery.

After some investigation I found out it was caused by some incompatibility of jQuery triggered events (```$(el).trigger("input")```) with Apline listeners. I am still not sure what is the difference between native and jQuery event that is causing this. We solved it by creating custom version of that jQuery plugin and replacit jQuery trigger with native ```el.dispatchEvent(new Event('input'))``` and now everything works.

I just wonder if there is any know way how to solve issue without creating and maintaining custom versions of that jQuery plugin? What is the reason jQuery events cannot be listened for by Alpine?

**Comment by SimoTod** [comment, marked as answer]
It's because jquery events are not standard ones, they are just a jQuery implementation to imitate the native event system. Alpine just use standard listeners under the hood (the one you could create with document.addEventListener).
Usually the options are:
1) you find a vanilla js version of that plugin
2) the jquery plugin offers hooks where you can dispatch standard events
3) you write something custom (like you did)

**Reply by MartinMystikJonas** [reply]
Ok we will go with option 3 for now and I will try other two.

It would be great if there would be some option how to make Apline and jQuery compatible some day. Maybe just option to redefine logic of adding event handlers from native to custom function - that can be set to use jQuery ```$(el).on()```.

**Reply by SimoTod** [reply]
It's a good idea actually. I don't think it would make it into the core because it would make Alpine dependent on Jquery and set a precedent but you could write your custom directive (basically you can copy x-on and change addEventListener with the Jquery counter part).

**Reply by MartinMystikJonas** [reply]
Dependency on jQuery is obviously unwanted. But I was thinking about allowing something like doing ```Alpine.setListenerAdder((el, event, handler, options) => $(el).on(event, handler, options))``` during alpine:init. This way there would be no dependency of Apine on jQuery but it would allow to simply solve compatibility issues when somebody use it both.

It seems that reimplementing x-on and x-model and also logic from utils/on.js just to solve this issue is a bit overkill and colud cause maintenence problems.

**Reply by MartinMystikJonas** [reply]
I did quick check how it could be implemented and it seems it would not be that hard. It could be similar to setEvaluator logic. I can try to draft a PR if you would be interested.

**Reply by SimoTod** [reply]
You need to ask Caleb since he's the maintener

**Reply by MartinMystikJonas** [reply]
@calebporzio What dou you think about this feature? I think I am not only one who would appreciate way to enable compatibility with jQuery events.

**Reply by MartinMystikJonas** [reply]
I drafted a PR here #2063

**Reply by SimoTod** [reply]
Ideally, you should add tests to avoid future regressions but also prove that the current critical features are still working (x-model, $dispatch, x-on itself).

**Reply by MartinMystikJonas** [reply]
I tried to draft tests as well but I am afraid it is beyond my skills.

I wrote it like this: 
```javascript
import { haveText, html, test } from '../utils'

test('can use jQuery events',
    html`
        <script src="https://code.jquery.com/jquery-3.6.0.slim.min.js"></script>
        <script>
            document.addEventListener('alpine:init', () => {
                Alpine.setListenerManipulators(
                    (target, event, handler, options, modifiers) => { $(target).on(event, handler) },
                    (target, event, handler, options) => { $(target).off(event, handler) }
                )
                $("span").trigger("click")
            })
        </script>

        <div x-data="{val: 'old'}">
            <span @click="val='new'" x-text="val"></span>
        </div>
    `,
    ({ get }) => get('span').should(haveText('new'))
)
```

But for some reason jQuery is not included and $ is undefined and I do not know what to do about it. Sorry.

But when I test my custom build manually in browser with our code it works
